pkg/storagev2/metrics.go

```diff
@@ -0,0 +1,320 @@
+	if err == nil {
+		m.metrics.IndexStoreCalls.WithLabelValues("Commit", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("Commit", "failure").Inc()
+	}
+	return err
+}
+
+// Rollback implements the TxStore interface.
+func (m txIndexStoreWithMetrics) Rollback() error {
+	now := time.Now()
+	err := m.TxStore.Rollback()
+	m.metrics.IndexStoreCallsDuration.WithLabelValues("Rollback").Observe(float64(time.Since(now).Nanoseconds()))
+	if err == nil {
+		m.metrics.IndexStoreCalls.WithLabelValues("Rollback", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("Rollback", "failure").Inc()
+	}
+	return err
+}
+
+// Close implements the TxStore interface.
+func (m txIndexStoreWithMetrics) Close() error {
+	now := time.Now()
+	err := m.TxStore.Close()
+	m.metrics.IndexStoreCallsDuration.WithLabelValues("Close").Observe(float64(time.Since(now).Nanoseconds()))
+	if err == nil {
+		m.metrics.IndexStoreCalls.WithLabelValues("Close", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("Close", "failure").Inc()
+	}
+	return err
+}
+
+// Get implements the TxStore interface.
+func (m txIndexStoreWithMetrics) Get(item Item) error {
+	now := time.Now()
+	err := m.TxStore.Get(item)
+	m.metrics.IndexStoreCallsDuration.WithLabelValues("Get").Observe(float64(time.Since(now).Nanoseconds()))
+	if err == nil || errors.Is(err, ErrNotFound) {
+		m.metrics.IndexStoreCalls.WithLabelValues("Get", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("Get", "failure").Inc()
+	}
+	return err
+}
+
+// Has implements the TxStore interface.
+func (m txIndexStoreWithMetrics) Has(key Key) (bool, error) {
+	now := time.Now()
+	has, err := m.TxStore.Has(key)
+	m.metrics.IndexStoreCallsDuration.WithLabelValues("Has").Observe(float64(time.Since(now).Nanoseconds()))
+	if err == nil || errors.Is(err, ErrNotFound) {
+		m.metrics.IndexStoreCalls.WithLabelValues("Has", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("Has", "failure").Inc()
+	}
+	return has, err
+}
+
+// GetSize implements the TxStore interface.
+func (m txIndexStoreWithMetrics) GetSize(key Key) (int, error) {
+	now := time.Now()
+	size, err := m.TxStore.GetSize(key)
+	m.metrics.IndexStoreCallsDuration.WithLabelValues("GetSize").Observe(float64(time.Since(now).Nanoseconds()))
+	if err == nil || errors.Is(err, ErrNotFound) {
+		m.metrics.IndexStoreCalls.WithLabelValues("GetSize", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("GetSize", "failure").Inc()
+	}
+	return size, err
+}
+
+// Iterate implements the TxStore interface.
+func (m txIndexStoreWithMetrics) Iterate(query Query, fn IterateFn) error {
+	now := time.Now()
+	err := m.TxStore.Iterate(query, fn)
+	m.metrics.IndexStoreCallsDuration.WithLabelValues("Iterate").Observe(float64(time.Since(now).Nanoseconds()))
+	if err == nil || errors.Is(err, ErrNotFound) {
```

Comment:
I think if we keep adding cases here for errors we will complicate this. I would prefer if we check for errors blindly at this level. We dont really know the context of the error so we should not infer stuff from it.

Same comment for all the methods.

Reply:
The `ErrNotFound` is a sentinel, so instead of returning another parameter, a boolean flag indicating that the record was/wasn't found we use this sentinel error. So counting it as an error would be a false positive.